lambda/enable_guardduty.py

```diff
@@ -0,0 +1,330 @@
+# Process documented here:
+# https://docs.aws.amazon.com/guardduty/latest/ug/guardduty_accounts.html#guardduty_become_api
+
+
+def handler(event, context):
+    '''
+    message = {
+        'account_id': 'string',
+        'message': 'string',  // optional, if un-specified, ENV var (from CFT)
+        'dry_run': true|false,  // optional, if un-specified, dry_run=false
+        'region': ['string'],  // optional, if un-specified, runs all regions
+    }
+    '''
+    logger.info("Received event: " + json.dumps(event, sort_keys=True))
+    message = json.loads(event['Records'][0]['Sns']['Message'])
+    logger.info("Received message: " + json.dumps(message, sort_keys=True))
+
+    if "account_id" not in message:
+        error_message = "message['account_id'] must be specified"
+        logger.error(error_message)
+        raise KeyError(error_message)
+
+    # get parent organization's account_id
+    role_arn = create_role_arn(message['account_id'], os.environ['ASSUME_ROLE'])
+    creds = get_creds(role_arn)
+    if creds is False:
+        logger.error(f"Unable to assume role {role_arn} to get parent organization")
+        return
+    org_client = boto3.client(
+        'organizations',
+        aws_access_key_id=creds['AccessKeyId'],
+        aws_secret_access_key=creds['SecretAccessKey'],
+        aws_session_token=creds['SessionToken'],
+    )
+    response = org_client.describe_organization()
+    message["payer_account_id"] = response['Organization']['MasterAccountId']
+    logger.info(f"Found payer account_id: {message['payer_account_id']}")
+
+    if "message" not in message:
+        logger.info(f"message['message'] not specified; default = {os.environ['INVITE_MESSAGE']}")
+
+    if "dry_run" not in message:
+        logger.info("message['dry_run'] not specified; default = False")
+        message['dry_run'] = False
+
+    if "region" not in message or not message["region"]:
+        logger.info("message['region'] not specified; default = all regions")
+        ec2 = boto3.client('ec2')
+        response = ec2.describe_regions()
+        message['region'] = [r['RegionName'] for r in response['Regions']]
+
+    for region in message['region']:
+        process_region(message, region)
+
+
+def process_region(event, region):
+    logger.info(f"Processing Region: {region}")
+
+    role_arn = create_role_arn(event["payer_account_id"], os.environ["PAYER_ROLE"])
+    creds = get_creds(role_arn)
+    if creds is False:
+        logger.error(f"Unable to assume role in payer {role_arn}")
+
+    gd_client = boto3.client(
+        'guardduty',
+        region_name=region,
+        aws_access_key_id=creds['AccessKeyId'],
+        aws_secret_access_key=creds['SecretAccessKey'],
+        aws_session_token=creds['SessionToken'],
+    )
+
+    try:
+        response = gd_client.list_detectors()
+        if len(response['DetectorIds']) == 0:
+            # We better create one
+            detector_id = create_parent_detector(gd_client, event, region)
```

Comment:
If I had named this create_masteraccount_detector() it would have been less confusing